integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java

```diff
@@ -5,61 +5,118 @@
 import org.apache.spark.Partition;
 import org.apache.spark.api.python.PythonRDD;
 import org.apache.spark.rdd.RDD;
 import org.apache.spark.sql.catalyst.analysis.FunctionRegistry;
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
 import org.apache.spark.sql.execution.SQLExecutionRDD;
-import org.apache.spark.sql.sources.BaseRelation;
-import scala.PartialFunction;
-import scala.runtime.AbstractPartialFunction;
 
 /**
  * {@link LogicalPlan} serializer which serialize {@link LogicalPlan} to JSON string. This
  * serializer relies on Jackson's {@link com.fasterxml.jackson.module.scala.DefaultScalaModule},
- * which is scala-version specific. Currently, we depend on the Scala 2.11 version of the jar,
- * making this code incompatible with Spark 3 or other other Spark installations compiled with Scala
- * 2.12 or higher. In such cases, we'll fail to load the Jackson module, but will continue to
- * attempt serialization of the {@link LogicalPlan}. If serialization fails, the exception message
- * and stacktrace will be reported.
+ * which is scala-version specific. That is why we need to use Jackson library available within
+ * Spark and not shaded one included within openlineage-java. This poses some limitations as logical
+ * plan serialization relies on mixIns and annotations, however annotations class are relocated and
+ * cannot be replaced with dynamic proxy classes. The major limitation of this approach is that we
+ * cannot use JsonTypeInfo annotation to enrich jsons with class names within the plan.
  */
 @Slf4j
 class LogicalPlanSerializer {
   private static final int MAX_SERIALIZED_PLAN_LENGTH =
       50000; // 50K UTF-8 chars should be ~200KB + some extra bytes added during json encoding
-  private final ObjectMapper mapper;
+  private final Object objectMapper;
+
+  /** relocate plugin rewrites by default all occurences of com.fasterxml.jackson */
+  private static final String UNSHADED_JACKSON_PACKAGE = "com.".trim() + "fasterxml.jackson";
 
   public LogicalPlanSerializer() {
-    mapper = new ObjectMapper();
+    // we need to use non-relocated Spark's ObjectMapper to serialize logical plan
+    objectMapper = getObjectMapper();
+
     try {
-      mapper.registerModule(
-          (Module)
-              Class.forName("com.fasterxml.jackson.module.scala.DefaultScalaModule$")
-                  .getDeclaredField("MODULE$")
-                  .get(null));
-    } catch (Exception t) {
-      log.warn("Can't register jackson scala module for serializing LogicalPlan");
+      Class clazz =
+          Class.forName(
+              UNSHADED_JACKSON_PACKAGE + ".databind.introspect.ClassIntrospector$MixInResolver");
+
+      /**
+       * Translation class which allows merging non-shaded Jackson available within Spark context
+       * with LogicalPlanMixinResolver that extends shaded Jackson MixinResolver from the shaded
+       * library.
+       */
+      Object resolver =
```

Comment:
Would be nice to explain more how it works or post link to documentation. I'm sure I'll scratch my head if I look at it in a year.